File: dsa_leetcode_beginner/0212-word-search-ii/0212-word-search-ii.js

class TrieNode {
    constructor() {
        this.children = {};
        this.isEnd = false;
    }
}

var findWords = function(board, words) {
    const root = buildTrie(words);
    const result = new Set();
    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j < board[0].length; j++) {
            if (root.children[board[i][j]]) {
                dfs(board, root, i, j, '', result);
            }
        }
    }
    return Array.from(result);
};

function buildTrie(words) {
    const root = new TrieNode();
    for (const word of words) {
        let node = root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        node.isEnd = true;
    }
    return root;
}

function dfs(board, node, i, j, path, result) {
    if (node.isEnd) {
        result.add(path);
        node.isEnd = false; // Avoid duplicates
    }
    if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || !node.children[board[i][j]]) {
        return;
    }
    const temp = board[i][j];
    board[i][j] = '#'; // Mark as visited
    for (const [x, y] of [[i+1, j], [i-1, j], [i, j+1], [i, j-1]]) {
        dfs(board, node.children[temp], x, y, path + temp, result);
    }
    board[i][j] = temp; // Unmark
}

File: dsa_leetcode_beginner/0212-word-search-ii/0212-word-search-ii.java

import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEnd = false;
}

public class WordSearchII {
    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = buildTrie(words);
        Set<String> result = new HashSet<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (root.children.containsKey(board[i][j])) {
                    dfs(board, root, i, j, "", result);
                }
            }
        }
        return new ArrayList<>(result);
    }

    private TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                node.children.putIfAbsent(c, new TrieNode());
                node = node.children.get(c);
            }
            node.isEnd = true;
        }
        return root;
    }

    private void dfs(char[][] board, TrieNode node, int i, int j, String path, Set<String> result) {
        if (node.isEnd) {
            result.add(path);
            node.isEnd = false; // Avoid duplicates
        }
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || !node.children.containsKey(board[i][j])) {
            return;
        }
        char temp = board[i][j];
        board[i][j] = '#'; // Mark as visited
        for (int[] dir : new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}) {
            int x = i + dir[0], y = j + dir[1];
            dfs(board, node.children.get(temp), x, y, path + temp, result);
        }
        board[i][j] = temp; // Unmark
    }
}
